client/my-sites/term-tree-selector/add-term.jsx

```diff
@@ -0,0 +1,231 @@
+import classNames from 'classnames';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { find } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Dialog from 'components/dialog';
+import TermTreeSelectorTerms from './terms';
+import Button from 'components/button';
+import Gridicon from 'components/gridicon';
+import FormInputValidation from 'components/forms/form-input-validation';
+import FormTextInput from 'components/forms/form-text-input';
+import FormSectionHeading from 'components/forms/form-section-heading';
+import FormCheckbox from 'components/forms/form-checkbox';
+import FormLabel from 'components/forms/form-label';
+import FormLegend from 'components/forms/form-legend';
+import FormFieldset from 'components/forms/form-fieldset';
+import viewport from 'lib/viewport';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getPostTypeTaxonomy } from 'state/post-types/taxonomies/selectors';
+import { getTerms } from 'state/terms/selectors';
+import { addTerm } from 'state/terms/actions';
+
+class TermSelectorAddTerm extends Component {
+	static initialState = {
+		showDialog: false,
+		selectedParent: [],
+		isTopLevel: true,
+		isValid: false,
+		error: null
+	};
+
+	static propTypes = {
+		labels: PropTypes.object,
+		postType: PropTypes.string,
+		siteId: PropTypes.number,
+		terms: PropTypes.array,
+		taxonomy: PropTypes.string,
+		translate: PropTypes.func
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = this.constructor.initialState;
+		this.boundCloseDialog = this.closeDialog.bind( this );
+		this.boundOpenDialog = this.openDialog.bind( this );
+		this.boundOnParentChange = this.onParentChange.bind( this );
+		this.boundOnSearch = this.onSearch.bind( this );
+		this.boundSaveTerm = this.saveTerm.bind( this );
+		this.boundOnTopLevelChange = this.onTopLevelChange.bind( this );
+		this.boundValidateInput = this.validateInput.bind( this );
+	}
+
+	onSearch( searchTerm ) {
+		this.setState( { searchTerm: searchTerm } );
+	}
+
+	openDialog( event ) {
+		event.preventDefault();
+
+		this.setState( {
+			showDialog: true,
+			selectedParent: []
+		} );
+	}
+
+	closeDialog() {
+		this.setState( this.constructor.initialState );
+	}
+
+	onParentChange( item ) {
+		this.setState( {
+			selectedParent: [ item.ID ],
+			isTopLevel: false
+		}, this.isValid );
+	}
+
+	onTopLevelChange() {
+		this.setState( {
+			isTopLevel: ! this.state.isTopLevel,
+			selectedParent: []
+		}, this.isValid );
+	}
+
+	getFormValues() {
+		const name = ReactDom.findDOMNode( this.refs.termName ).value.trim();
+		const parent = this.state.selectedParent.length ? this.state.selectedParent[ 0 ] : 0;
+
+		return { name, parent };
+	}
+
+	isValid() {
+		let error;
+
+		const values = this.getFormValues();
+
+		if ( ! values.name.length ) {
+			error = true;
+		}
+
+		const lowerCasedTermName = values.name.toLowerCase();
+		const matchingTerm = find( this.props.terms, ( term ) => {
```

Comment:
No need for change, but [`Array#some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) / [`_.some`](http://lodash.com/docs#some) is more to the point of what your use-case is here (boolean return value if matching found, instead of full object).